src/TreeSitter/Importing.hs

```diff
@@ -1,116 +1,146 @@
+             ScopedTypeVariables, TypeApplications, TypeOperators #-}
+module TreeSitter.Importing
+( parseByteString
+, FieldName(..)
+, Building(..)
+, step
+, push
+, goto
+, peekNode
+, peekFieldName
+, getFields
+) where
+
+import           Control.Applicative
+import           Control.Effect hiding ((:+:))
+import           Control.Effect.Reader
+import           Control.Monad (void)
+import           Control.Monad.IO.Class
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString as B
+import qualified Data.Map as Map
+import qualified Data.Text as Text
+import           Data.Text.Encoding
+import           Foreign.C.String
 import           Foreign.Marshal.Alloc
+import           Foreign.Marshal.Utils
 import           Foreign.Ptr
 import           Foreign.Storable
-import TreeSitter.Cursor as TS
-import TreeSitter.Node as TS
-import TreeSitter.Parser as TS
-import TreeSitter.Tree as TS
-import qualified Data.Text as Text
-import Control.Effect
-import Control.Effect.Reader
-import Control.Monad.IO.Class
-import Data.Text.Encoding
-import qualified Data.ByteString as B
-import Control.Applicative
-import Control.Monad (void)
-
-data Expression
-      = NumberExpression Number | IdentifierExpression Identifier
-      deriving (Eq, Ord, Show)
-
-data Number = Number
-      deriving (Eq, Ord, Show)
-
-data Identifier = Identifier
-      deriving (Eq, Ord, Show)
-
-
-importByteString :: (Importing t) => Ptr TS.Parser -> ByteString -> IO (Maybe t)
-importByteString parser bytestring =
-  unsafeUseAsCStringLen bytestring $ \ (source, len) -> alloca (\ rootPtr -> do
-      let acquire =
-            ts_parser_parse_string parser nullPtr source len
-
-      let release t
-            | t == nullPtr = pure ()
-            | otherwise = ts_tree_delete t
-
-      let go treePtr =
-            if treePtr == nullPtr
-              then pure Nothing
-              else do
-                ts_tree_root_node_p treePtr rootPtr
-                withCursor (castPtr rootPtr) $ \ cursor ->
-                  Just <$> runM (runReader cursor (runReader bytestring import'))
-      Exc.bracket acquire release go)
-
-withCursor :: Ptr TSNode -> (Ptr Cursor -> IO a) -> IO a
-withCursor rootPtr action = allocaBytes sizeOfCursor $ \ cursor -> Exc.bracket_
-  (ts_tree_cursor_new_p rootPtr cursor)
-  (ts_tree_cursor_delete cursor)
-  (action cursor)
-
-instance Importing a => Importing [a] where
-  import' = push $ do
-    a <- import' @a
-    pure [a]
-
-
-instance (Importing a, Importing b) => Importing (a,b) where
-  import' = push $ do
-    a <- import' @a
-    step
-    b <- import' @b
-    pure (a, b)
+import           GHC.Generics
+import           TreeSitter.Cursor as TS
+import           TreeSitter.Language as TS
+import           TreeSitter.Node as TS
+import           TreeSitter.Parser as TS
+import           TreeSitter.Tree as TS
+
+parseByteString :: Building t => Ptr TS.Language -> ByteString -> IO (Maybe t)
```

Comment:
We take source code in the form of a bytestring and produce AST values represented by `IO Maybe`